File: mParticle-Apple-SDK/Utils/MPConvertJS.m
#import "MPConvertJS.h"
#import "MPCommerceEvent+Dictionary.h"
#import "mParticle.h"
#import "MPILogger.h"
@import mParticle_Apple_SDK_Swift;

@implementation MPConvertJS_PRIVATE

+ (MPCommerceEventAction)commerceEventAction:(NSNumber *)json {
    NSNumber *action = [MPConvertJSFields commerceEventActionForJSValue:json];
    if (action == nil) {
        MPILogError(@"Invalid commerce event action received from webview: %@", json);
        return MPCommerceEventActionAddToCart;
    }

    return (MPCommerceEventAction)action.unsignedIntegerValue;
}

+ (MPCommerceEvent *)commerceEvent:(NSDictionary *)json {
    if (json[@"ProductAction"] != nil && ![json[@"ProductAction"] isKindOfClass:[NSDictionary class]]) {
        MPILogError(@"Unexpected commerce event data received from webview");
        return nil;
    }
    
    NSDictionary *productAction = json[@"ProductAction"];
    BOOL isProductAction = productAction[@"ProductActionType"] != nil;
    BOOL isPromotion = json[@"PromotionAction"] != nil;
    BOOL isImpression = json[@"ProductImpressions"] != nil;
    BOOL isValid = isProductAction || isPromotion || isImpression;

    if (!isValid) {
        MPILogError(@"Invalid commerce event dictionary received from webview: %@", json);
        return nil;
    }

    MPCommerceEvent *commerceEvent = nil;
    if (isProductAction) {
        id productActionType = productAction[@"ProductActionType"];
        if (!productActionType || ![productActionType isKindOfClass:[NSNumber class]]) {
            MPILogError(@"Unexpected product action type received from webview");
            return nil;
        }
        MPCommerceEventAction action = [MPConvertJS_PRIVATE commerceEventAction:productActionType];
        commerceEvent = [[MPCommerceEvent alloc] initWithAction:action];
    }
    else if (isPromotion) {
        MPPromotionContainer *promotionContainer = [MPConvertJS_PRIVATE promotionContainer:json];
        commerceEvent = [[MPCommerceEvent alloc] initWithPromotionContainer:promotionContainer];
    }
    else {
        commerceEvent = [[MPCommerceEvent alloc] initWithImpressionName:nil product:nil];
    }
    
    id eventAttributes = json[@"EventAttributes"];
    if ([eventAttributes isKindOfClass:[NSDictionary class]]) {
        commerceEvent.customAttributes = (NSDictionary *)eventAttributes;
    }
    
    id checkoutOptionsObj = json[@"CheckoutOptions"];
    if ([checkoutOptionsObj isKindOfClass:[NSString class]]) {
        commerceEvent.checkoutOptions = (NSString *)checkoutOptionsObj;
    }

    id productActionListNameObj = json[@"productActionListName"];
    if ([productActionListNameObj isKindOfClass:[NSString class]]) {
        commerceEvent.productListName = (NSString *)productActionListNameObj;
    }

    id productActionListSourceObj = json[@"productActionListSource"];
    if ([productActionListSourceObj isKindOfClass:[NSString class]]) {
        commerceEvent.productListSource = (NSString *)productActionListSourceObj;
    }

    id currencyCodeObj = json[@"CurrencyCode"];
    if ([currencyCodeObj isKindOfClass:[NSString class]]) {
        commerceEvent.currency = (NSString *)currencyCodeObj;
    }
    
    if (productAction != nil) {
        commerceEvent.transactionAttributes = [self transactionAttributes:productAction];
    }
    
    id checkoutStepObj = json[@"CheckoutStep"];
    if ([checkoutStepObj isKindOfClass:[NSNumber class]]) {
        commerceEvent.checkoutStep = [(NSNumber *)checkoutStepObj intValue];
    }
    
    id customFlagsObj = json[@"CustomFlags"];
    if ([customFlagsObj isKindOfClass:[NSDictionary class]]) {
        NSDictionary *customFlags = (NSDictionary *)customFlagsObj;

        for (id key in customFlags) {
            id value = customFlags[key];

            if ([value isKindOfClass:[NSArray class]]) {
                BOOL allStrings = YES;
                for (id item in (NSArray *)value) {
                    if (![item isKindOfClass:[NSString class]]) { allStrings = NO; break; }
                }
                if (allStrings && [key isKindOfClass:[NSString class]]) {
                    [commerceEvent addCustomFlags:(NSArray<NSString *> *)value withKey:(NSString *)key];
                }

            } else if ([value isKindOfClass:[NSString class]]) {
                if ([key isKindOfClass:[NSString class]]) {
                    [commerceEvent addCustomFlag:(NSString *)value withKey:(NSString *)key];
                }
            }
        }
    }

    id productListObj = productAction[@"ProductList"];
    if ([productListObj isKindOfClass:[NSArray class]]) {
        NSArray *jsonProducts = (NSArray *)productListObj;

        NSMutableArray<MPProduct *> *products = [NSMutableArray arrayWithCapacity:jsonProducts.count];
        for (id item in jsonProducts) {
            if (![item isKindOfClass:[NSDictionary class]]) { continue; }

            MPProduct *p = [self product:(NSDictionary *)item];
            if (p) { [products addObject:p]; }
        }
        [commerceEvent addProducts:products];
    }

    id impressionsObj = json[@"ProductImpressions"];
    if ([impressionsObj isKindOfClass:[NSArray class]]) {
        NSArray *jsonImpressions = (NSArray *)impressionsObj;

        for (id impressionItem in jsonImpressions) {
            if (![impressionItem isKindOfClass:[NSDictionary class]]) { continue; }

            NSDictionary *jsonImpression = (NSDictionary *)impressionItem;
            id listNameObj = jsonImpression[@"ProductImpressionList"];
            id impressionProductsObj = jsonImpression[@"ProductList"];

            if ([listNameObj isKindOfClass:[NSString class]] &&
                [impressionProductsObj isKindOfClass:[NSArray class]]) {

                NSString *listName = (NSString *)listNameObj;
                NSArray *impressionProducts = (NSArray *)impressionProductsObj;

                for (id prodItem in impressionProducts) {
                    if (![prodItem isKindOfClass:[NSDictionary class]]) { continue; }

                    MPProduct *product = [MPConvertJS_PRIVATE product:(NSDictionary *)prodItem];
                    [commerceEvent addImpression:product listName:listName];
                }
            }
        }
    }

    return commerceEvent;
}

+ (MPPromotionContainer *)promotionContainer:(NSDictionary *)json {
    NSDictionary *promotionActionDictionary = json[@"PromotionAction"];
    if (!promotionActionDictionary || ![promotionActionDictionary isKindOfClass:[NSDictionary class]]) {
        MPILogError(@"Unexpected promotion container action data received from webview");
        return nil;
    }
    
    NSNumber *promotionActionTypeNumber = promotionActionDictionary[@"PromotionActionType"];
    if (promotionActionTypeNumber == nil || ![promotionActionTypeNumber isKindOfClass:[NSNumber class]]) {
        MPILogError(@"Unexpected promotion container action type data received from webview");
        return nil;
    }
    
    int promotionActionInt = [promotionActionTypeNumber intValue];
    MPPromotionAction promotionAction = promotionActionInt == 1 ? MPPromotionActionView : MPPromotionActionClick;
    MPPromotionContainer *promotionContainer = [[MPPromotionContainer alloc] initWithAction:promotionAction promotion:nil];
    
    NSArray *jsonPromotions = promotionActionDictionary[@"PromotionList"];
    if (!jsonPromotions || ![jsonPromotions isKindOfClass:[NSArray class]]) {
        MPILogError(@"Unexpected promotion container list data received from webview");
        return nil;
    }
    
    [jsonPromotions enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        MPPromotion *promotion = [MPConvertJS_PRIVATE promotion:obj];
        [promotionContainer addPromotion:promotion];
    }];

    return promotionContainer;
}

+ (MPPromotion *)promotion:(NSDictionary *)json {
    MPPromotion *promotion = [[MPPromotion alloc] init];
    MPPromotionFieldsJS *fields = [MPConvertJSFields promotionFieldsFromJSON:json];

    // Guarded, not unconditional: -attributes/-beautifiedAttributes are lazy and
    // allocate on first touch, including from a setter's nil-removal branch. An
    // unconditional assignment when every field is nil still allocates an empty
    // (non-nil) _attributes, which -dictionaryRepresentation returns as-is —
    // turning "no valid fields" into a promotion MPPromotionContainer no longer
    // skips, instead of leaving it nil as before this file moved to Swift.
    if (fields.creative) { promotion.creative = fields.creative; }
    if (fields.name) { promotion.name = fields.name; }
    if (fields.position) { promotion.position = fields.position; }
    if (fields.promotionId) { promotion.promotionId = fields.promotionId; }

    return promotion;
}

+ (MPTransactionAttributes *)transactionAttributes:(NSDictionary *)json {
    MPTransactionAttributes *transactionAttributes = [[MPTransactionAttributes alloc] init];
    MPTransactionAttributesFieldsJS *fields = [MPConvertJSFields transactionAttributesFieldsFromJSON:json];

    // Same guard, same reason: MPTransactionAttributes has the identical lazy
    // -attributes/-beautifiedAttributes pattern as MPPromotion.
    if (fields.affiliation) { transactionAttributes.affiliation = fields.affiliation; }
    if (fields.couponCode) { transactionAttributes.couponCode = fields.couponCode; }
    if (fields.shipping) { transactionAttributes.shipping = fields.shipping; }
    if (fields.tax) { transactionAttributes.tax = fields.tax; }
    if (fields.revenue) { transactionAttributes.revenue = fields.revenue; }
    if (fields.transactionId) { transactionAttributes.transactionId = fields.transactionId; }

    return transactionAttributes;
}

+ (MPProduct *)product:(NSDictionary *)json {
    MPProduct *product = [[MPProduct alloc] init];

    id brand = json[@"Brand"];
    if ([brand isKindOfClass:[NSString class]]) {
        product.brand = (NSString *)brand;
    }

    id category = json[@"Category"];
    if ([category isKindOfClass:[NSString class]]) {
        product.category = (NSString *)category;
    }

    id couponCode = json[@"CouponCode"];
    if ([couponCode isKindOfClass:[NSString class]]) {
        product.couponCode = (NSString *)couponCode;
    }

    id name = json[@"Name"];
    if ([name isKindOfClass:[NSString class]]) {
        product.name = (NSString *)name;
    }

    id price = json[@"Price"];
    if ([price isKindOfClass:[NSNumber class]]) {
        product.price = (NSNumber *)price;
    } else if ([price isKindOfClass:[NSString class]]) {
        product.price = @([(NSString *)price doubleValue]);
    }

    id sku = json[@"Sku"];
    if ([sku isKindOfClass:[NSString class]]) {
        product.sku = (NSString *)sku;
    }

    id variant = json[@"Variant"];
    if ([variant isKindOfClass:[NSString class]]) {
        product.variant = (NSString *)variant;
    }

    id position = json[@"Position"];
    if ([position isKindOfClass:[NSNumber class]]) {
        product.position = [(NSNumber *)position unsignedIntValue];
    }

    id quantity = json[@"Quantity"];
    if ([quantity isKindOfClass:[NSNumber class]]) {
        product.quantity = (NSNumber *)quantity;
    }

    id attributes = json[@"Attributes"];
    if ([attributes isKindOfClass:[NSDictionary class]]) {
        NSDictionary *jsonAttributes = (NSDictionary *)attributes;
        [jsonAttributes enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
            if ([key isKindOfClass:[NSString class]] && [obj isKindOfClass:[NSString class]]) {
                [product setValue:obj forKey:key];
            }
        }];
    }

    return product;
}

+ (MPIdentityApiRequest *)identityApiRequest:(NSDictionary *)json {
    MPIdentityApiRequest *request = [MPIdentityApiRequest requestWithEmptyUser];
    
    NSArray *userIdentities = json[@"UserIdentities"];
    if (!userIdentities || ![userIdentities isKindOfClass:[NSArray class]]) {
        MPILogError(@"Unexpected user identity data received from webview");
        return nil;
    }
    
    if (userIdentities.count > 0) {
        __block BOOL allSuccess = YES;
        
        [userIdentities enumerateObjectsUsingBlock:^(NSDictionary * _Nonnull identityDictionary, NSUInteger idx, BOOL * _Nonnull stop) {
            NSString *identity = identityDictionary[@"Identity"];
            NSNumber *identityTypeNumber = identityDictionary[@"Type"];
            
            if (!identity || ![identity isKindOfClass:[NSString class]]) {
                allSuccess = NO;
                *stop = YES;
                return;
            }
            
            if (identityTypeNumber == nil || ![identityTypeNumber isKindOfClass:[NSNumber class]]) {
                allSuccess = NO;
                *stop = YES;
                return;
            }
            
            MPIdentity identityType = (MPIdentity)[identityTypeNumber unsignedIntegerValue];
            [request setIdentity:identity identityType:identityType];
        }];
        
        if (!allSuccess) {
            return nil;
        }
    }
    
    NSString *identity = json[@"Identity"];
    NSNumber *identityTypeNumber = json[@"Type"];
    
    if (identity && [identity isKindOfClass:[NSString class]] && 
        identityTypeNumber && [identityTypeNumber isKindOfClass:[NSNumber class]]) {
        MPIdentity identityType = (MPIdentity)[identityTypeNumber unsignedIntegerValue];
        [request setIdentity:identity identityType:identityType];
    }
    
    return request;
}

@end

File: mParticle-Apple-SDK/Utils/MPConvertJS.swift
//
//  MPConvertJS.swift
//  mParticle-Apple-SDK
//
//  Created by Brandon Stalnaker on 10/3/24.
//

import Foundation

@objc public enum MPJSCommerceEventAction: UInt, @unchecked Sendable {
    case unknown = 0
    case addToCart = 1
    case removeFromCart = 2
    case checkout = 3
    case checkoutOptions = 4
    case click = 5
    case viewDetail = 6
    case purchase = 7
    case refund = 8
    case addToWishList = 9
    case removeFromWishlist = 10
}

@objc final public class MPConvertJS_PRIVATE: NSObject {
    
    @objc public static func commerceEventAction(_ json: NSNumber) -> MPCommerceEventAction {
        switch json.uintValue {
        case MPJSCommerceEventAction.addToCart.rawValue:
            return .addToCart
        case MPJSCommerceEventAction.removeFromCart.rawValue:
            return .removeFromCart
        case MPJSCommerceEventAction.checkout.rawValue:
            return .checkout
        case MPJSCommerceEventAction.checkoutOptions.rawValue:
            return .checkoutOptions
        case MPJSCommerceEventAction.click.rawValue:
            return .click
        case MPJSCommerceEventAction.viewDetail.rawValue:
            return .viewDetail
        case MPJSCommerceEventAction.purchase.rawValue:
            return .purchase
        case MPJSCommerceEventAction.refund.rawValue:
            return .refund
        case MPJSCommerceEventAction.addToWishList.rawValue:
            return .addToWishList
        case MPJSCommerceEventAction.removeFromWishlist.rawValue:
            return .removeFromWishlist
        default:
            MPLog.error("Invalid commerce event action received from webview: %@", json)
            return .addToCart
        }
    }
    
    @objc public static func commerceEvent(_ json: [AnyHashable : Any]) -> MPCommerceEvent? {
        guard json["ProductAction"] == nil || json["ProductAction"] is [String: Any] else {
            MPLog.error("Unexpected commerce event data received from webview")
            return nil
        }
        
        let productAction = json["ProductAction"] as? [String: Any]
        let isProductAction = productAction?["ProductActionType"] != nil
        let isPromotion = json["PromotionAction"] != nil
        let isImpression = json["ProductImpressions"] != nil
        let isValid = isProductAction || isPromotion || isImpression
        
        if !isValid {
            MPLog.error("Invalid commerce event dictionary received from webview: %@", json)
            return nil
        }
        
        let commerceEvent: MPCommerceEvent
        if isProductAction {
            guard let productActionType = productAction?["ProductActionType"] as? NSNumber else {
                MPLog.error("Unexpected product action type received from webview")
                return nil
            }
            let action = Self.commerceEventAction(productActionType)
            commerceEvent = MPCommerceEvent(action: action)
        } else if isPromotion {
            let promotionContainer = Self.promotionContainer(json)
            commerceEvent = MPCommerceEvent(promotionContainer: promotionContainer)
        } else {
            commerceEvent = MPCommerceEvent(impressionName: nil, product: nil)
        }
        
        if let eventAttributes = json["EventAttributes"] as? [String : Any] {
            commerceEvent.customAttributes = eventAttributes
        }
        if let checkoutOptions = json["CheckoutOptions"] as? String {
            commerceEvent.checkoutOptions = checkoutOptions
        }
        if let productActionListName = json["productActionListName"] as? String {
            commerceEvent.productListName = productActionListName
        }
        if let productActionListSource = json["productActionListSource"] as? String {
            commerceEvent.productListSource = productActionListSource
        }
        if let currencyCode = json["CurrencyCode"] as? String {
            commerceEvent.currency = currencyCode
        }
        if let productAction = productAction {
            commerceEvent.transactionAttributes = Self.transactionAttributes(productAction)
        }
        if let checkoutStep = json["CheckoutStep"] as? NSNumber {
            commerceEvent.checkoutStep = checkoutStep.intValue
        }
        if let customFlags = json["CustomFlags"] as? [String : Any] {
            for key in customFlags.keys {
                if let valueArray = customFlags[key] as? [String] {
                    commerceEvent.addCustomFlags(valueArray, withKey: key)
                } else if let valueString = customFlags[key] as? String {
                    commerceEvent.addCustomFlag(valueString, withKey: key)
                }
            }
        }

        if let jsonProducts = productAction?["ProductList"] as? [[AnyHashable : Any]] {
            let products = jsonProducts.map { Self.product($0) }
            commerceEvent.addProducts(products)
        }

        if let jsonImpressions = json["ProductImpressions"] as? [[AnyHashable : Any]] {
            for jsonImpression in jsonImpressions {
                if let listName = jsonImpression["ProductImpressionList"] as? String,
                   let jsonProducts = jsonImpression["ProductList"] as? [[AnyHashable : Any]] {
                    for jsonObject in jsonProducts {
                        let product = MPConvertJS_PRIVATE.product(jsonObject)
                        commerceEvent.addImpression(product, listName: listName)
                    }
                }
            }
        }
        
        return commerceEvent
    }
    
    @objc public static func promotionContainer(_ json: [AnyHashable : Any]) -> MPPromotionContainer? {
        guard let promotionDictionary = (json["PromotionAction"] as? [String: Any]) else {
            MPLog.error("Unexpected promotion container action data received from webview")
            return nil
        }
        
        guard let promotionActionTypeNumber = (promotionDictionary["PromotionActionType"] as? NSNumber) else {
            MPLog.error("Unexpected promotion container action type data received from webview")
            return nil
        }
        
        let promotionAction = promotionActionTypeNumber.intValue == 1 ? MPPromotionAction.view : MPPromotionAction.click
        let promotionContainer = MPPromotionContainer(action: promotionAction, promotion: nil)
        if let jsonPromotions = promotionDictionary["PromotionList"] as? [[AnyHashable : Any]] {
            for jsonObject in jsonPromotions {
                promotionContainer.addPromotion(MPConvertJS_PRIVATE.promotion(jsonObject))
            }
        } else {
            MPLog.error("Unexpected promotion container list data received from webview")
            return nil
        }

        return promotionContainer;
    }
    
    @objc public static func promotion(_ json: [AnyHashable : Any]) -> MPPromotion {
        let promotion = MPPromotion()

        if let creative = json["Creative"] as? String {
            promotion.creative = creative
        }
        if let name = json["Name"] as? String {
            promotion.name = name
        }
        if let position = json["Position"] as? String {
            promotion.position = position
        }
        if let id = json["Id"] as? String {
            promotion.promotionId = id
        }
        
        return promotion;
    }
    
    @objc public static func transactionAttributes(_ json: [AnyHashable : Any] = [:]) -> MPTransactionAttributes! {
        let transactionAttributes = MPTransactionAttributes()
        
        if let affiliation = json["Affiliation"] as? String {
            transactionAttributes.affiliation = affiliation
        }
        if let couponCode = json["CouponCode"] as? String {
            transactionAttributes.couponCode = couponCode
        }
        if let shippingAmount = json["ShippingAmount"] as? NSNumber {
            transactionAttributes.shipping = shippingAmount
        }
        if let taxAmount = json["TaxAmount"] as? NSNumber {
            transactionAttributes.tax = taxAmount
        }
        if let totalAmount = json["TotalAmount"] as? NSNumber {
            transactionAttributes.revenue = totalAmount
        }
        if let transactionId = json["TransactionId"] as? String {
            transactionAttributes.transactionId = transactionId
        }
        
        return transactionAttributes;
    }
    
    @objc public static func product(_ json: [AnyHashable : Any]) -> MPProduct {
        let product = MPProduct()
        
        if let brand = json["Brand"] as? String {
            product.brand = brand
        }
        if let category = json["Category"] as? String {
            product.category = category
        }
        if let couponCode = json["CouponCode"] as? String {
            product.couponCode = couponCode
        }
        if let name = json["Name"] as? String {
            product.name = name
        }
        if let price = json["Price"] as? NSNumber {
            product.price = price
        } else if let price = json["Price"] as? String, let double = double_t(price) {
            product.price = NSNumber(value: double)
        }
        if let sku = json["Sku"] as? String {
            product.sku = sku
        }
        if let variant = json["Variant"] as? String {
            product.variant = variant
        }
        if let position = json["Position"] as? NSNumber {
            product.position = position.uintValue
        }
        if let quantity = json["Quantity"] as? NSNumber {
            product.quantity = quantity
        }
        if let jsonAttributes = json["Attributes"] as? [String : String] {
            for (key, value) in jsonAttributes {
                product.setValue(value, forKey: key)
            }
        }

        return product;
    }
    
    @objc private static func identityFromNumber(_ identityTypeNumber: NSNumber) -> MPIdentity {
        return MPIdentity(rawValue: identityTypeNumber.uintValue) ?? .other
    }
    
    @objc public static func identityApiRequest(_ json: [AnyHashable : Any]?) -> MPIdentityApiRequest? {
        let request = MPIdentityApiRequest.withEmptyUser()
        
        guard let userIdentities = json?["UserIdentities"] as? [[AnyHashable : Any]] else {
            MPLog.error("Unexpected user identity data received from webview")
            return nil
        }
        
        for identityDictionary in userIdentities {
            if let identity = identityDictionary["Identity"] as? String,
               let identityTypeNumber = identityDictionary["Type"] as? NSNumber,
               let identityType = MPIdentity(rawValue: identityTypeNumber.uintValue) {
                request.setIdentity(identity, identityType: identityType)
            } else {
                return nil
            }
        }

        if let identity = json?["Identity"] as? String,
           let identityTypeNumber = json?["Type"] as? NSNumber,
           let identityType = MPIdentity(rawValue: identityTypeNumber.uintValue) {
            request.setIdentity(identity, identityType: identityType)
        }
        
        return request
    }
}
